service/src/event/service.rs

```diff
@@ -197,225 +177,88 @@ impl CeramicEventService {
-            } else {
-                let init_cid = insertable.body.cid;
-                result.mark_event_deliverable_now(insertable, init_cid);
-            }
-        }
+        let ordered = OrderEvents::try_new(&self.pool, to_insert).await?;
 
-        trace!(local_events_checking=%insert_after_history_check.len(), "checking local history");
-        result
-            .verify_history_inline(pool, insert_after_history_check)
-            .await?;
-        Ok(result)
-    }
+        let missing_history = ordered
+            .missing_history
+            .iter()
+            .map(|(e, _)| e.order_key.clone())
+            .collect();
 
-    async fn parse_item<'a>(
-        item: &ReconItem<'a, EventId>,
-    ) -> Result<(EventInsertable, Option<DeliverableMetadata>)> {
-        let cid = item.key.cid().ok_or_else(|| {
-            Error::new_invalid_arg(anyhow::anyhow!("EventID is missing a CID: {}", item.key))
-        })?;
-        // we want to end a conversation if any of the events aren't ceramic events and not store them
-        // this includes making sure the key matched the body cid
-        let (insertable_body, maybe_prev) =
-            CeramicEventService::parse_event_carfile(cid, item.value).await?;
-        let insertable = EventInsertable::try_new(item.key.to_owned(), insertable_body)?;
-        Ok((insertable, maybe_prev))
-    }
+        let to_insert_with_metadata = if history_required {
+            ordered.deliverable
+        } else {
+            ordered
+                .deliverable
+                .into_iter()
+                .chain(ordered.missing_history)
+                .collect()
+        };
 
-    fn mark_event_deliverable_later(
-        &mut self,
-        insertable: EventInsertable,
-        meta: DeliverableMetadata,
-    ) {
-        self.background_task_deliverable
-            .insert(insertable.body.cid, meta);
-        self.insert_now.push(insertable);
-    }
+        let to_insert = to_insert_with_metadata
+            .iter()
+            .map(|(e, _)| e.clone())
+            .collect::<Vec<_>>();
 
-    fn mark_event_deliverable_now(&mut self, mut ev: EventInsertable, init_cid: Cid) {
-        ev.set_deliverable(true);
-        self.notify_task_new
-            .push(DeliveredEvent::new(ev.body.cid, init_cid));
-        self.insert_now.push(ev);
-    }
+        let res = CeramicOneEvent::insert_many(&self.pool, &to_insert[..]).await?;
 
-    async fn verify_history_inline(
-        &mut self,
-        pool: &SqlitePool,
-        to_check: Vec<(DeliverableMetadata, EventInsertable)>,
-    ) -> Result<()> {
-        if to_check.is_empty() {
-            return Ok(());
+        // api writes shouldn't have any missed pieces that need ordering so we don't send those and return early
+        if history_required {
+            return Ok(InsertResult {
+                store_result: res,
+                missing_history,
+            });
```

Comment:
nit, can you remove this early return and guard the discovered logic with `!history_required`?

Reply:
I'm pretty sure I asked for the exact opposite in an earlier review 😅.

Can I ask why you don't like the early return?  Personally I'm a big fan of early returns, I like removing the extra indentation layer and I think it helps reduce cognitive load by letting me see that this case is over and dealt with and I don't need to think about it any more, vs being in a big if block where there's some overhead to remembering what block I'm in and what the condition in play is right now.

Reply:
I generally dislike early returns as they make it harder to determine the different code branches as they can be easy to miss. In this case both returns, return the same value so they were really the same code branch just duplicated.   

Not a strong preference here either way.

Reply:
fascinating!  I've never heard an argument against early returns before.  I often push for early returns when I do code reviews because I've always felt they improve code readability.  I always thought that opinion was non-controversial 😅

Good to know it's not as clear-cut as I thought.  I still like them but don't feel strongly either way.  Maybe I'll stop pushing as hard for others to use them going forward

Reply:
I also don't like early returns unless they're right away. At least in rust I prefer to rely on the fact that it's an expression and returns something. 

That said, I think I can extract the big if block so it's easy to understand (and hopefully satisfies everyone 😁).